src/coreclr/vm/finalizerthread.cpp

```diff
@@ -258,6 +259,14 @@ VOID FinalizerThread::FinalizerThreadWorker(void *args)
 
         WaitForFinalizerEvent (hEventFinalizer);
 
+        // We can potentially aggregate multiple requests to do work on finalizer action here
```

Comment:
This is called in pre-emptive mode. It means that the GC can be running in parallel with this and modifying the finalizer action. It sounds like a race condiiton.

Reply:
In an unlikely event, it is possible that the finalizer thread is parked right after `WaitForFinalizerEvent` and before `GetFinalizerAction`. The NoGCRegion goes on allocate and reaches the threshold, then then this sequence of events could happen.

1. `GetFinalizerAction` reads `current_finalizer_action_kind` and gets `finalizer_action_kind::none`
2. GC sets the `current_finalizer_action_kind` to `(finalizer_action_kind::execute_no_gc_callback | finalizer_action_kind::detach_no_gc_callback)`
3. `GetFinalizerAction` write `current_finalizer_action_kind` to `finalizer_action_kind::none`
4. GC calls `GCToEEInterface::EnableFinalization(true);`, we triggered the finalization thread. The event is set again.
5. The finalizer thread moves on and run all the finalizers without the callback, and it loops back, the event is set, but the `GetFinalizerAction` will return `false`, so we missed calling the callback.

IMO, this is very unlikely. How could the finalizer thread wait for that long at that spot? But anyway, this should be fixable by changing the `GetFinalizerAction` to use `InterlockedExchange`. With that, it is impossible for step 2 to slip between 1 and 3. Either step 2 happens before (1,3) or step 2 after (1,3), we will not miss the callback.

What else do you think might go wrong?

Reply:
> IMO, this is very unlikely. How could the finalizer thread wait for that long at that spot? 

It can happen very easily on overloaded machine.

Reply:
> What else do you think might go wrong?

You may also want to think about the side that sets the finalizer_action static variable. The Interlocked.Exchange can be executed between reading and writing of `current_finalizer_action_kind` in `current_finalizer_action_kind |= finalizer_action_kind::detach_no_gc_callback;`.

Reply:
The latest version should have the concurrency issue related to `current_finalizer_action_kind` fixed for all possible races I know of.

Here are all the operations that could change `current_finalizer_action_kind`.

Op 1: Enabling a callback [set callback_id and then set the action to none]
Op 2: Firing a callback [set the action to execute|detach atomically]
Op 3: Ending a NoGCRegion [interlocked or set the action to action|detach]
Op 4: Finalizer picking up work [interlocked exchange the action with 0]

By definition, only Op4 can race with the other 3 operations because we would have blocked the operation because the system is not in the right state to begin with.

The case for the race between Op4 and (Op2 or Op3) are similar. In these cases, either the finalizer wins or not:

In case the finalizer wins, the work is set later, but then we also the event, so the event is on and the work will be picked up in the next round.

The `callback_id` is added to make sure doing work in the next round is the right thing to do. If the finalizer thread got the work, but then sleep for indefinite amount of time before it actually does the work, the `callback_id` mismatch will cause it to drop the task and not do the wrong thing.

In case the GC wins, the finalizer will pick up the work and reset it right away, so the work is also done.  

The case for the race between Op4 and Op1 is interesting, and there can be 3 cases:

Case 1: Finalizer wins - the finalizer pick up the old work and did it. This is okay because we didn't promise we will cancel the callback. This does mean the user code need to be careful that a late callback might cancel the old game, but we already discussed that case.

Case 2: GC set action first, the finalizer pick up a no-op and did nothing, and GC set callback ID later.

Case 3: GC set action first, GC set callback ID later, and the finalizer pickup a no-op work did nothing.

The order is important, if GC set callback_id first, the finalizer will pick up the wrong work and that's no good. Order is guaranteed by the use of volatile so we should be fine.

That concludes the analysis for concurrency correctness. 

Do we see other possibilities for things to go wrong?

Reply:
> What about if the thread is rescheduled right after we check the cookie, but before actually doing the work?

In the new iteration, I addressed this possible race condition by introducing `callbackWithId`.

The key idea is that when we detach the callback, we will **not** set the `callbackWithId` to null, so there is no possibility that the finalizer inadvertently detachs the callback for the next NoGCRegion.

The change leads to the possibility that `callbackWithId` is not null but `callbackWithId.Callback` is null. That happens precisely after the callback is detached by the finalizer. At this point, it is possible (and supported) that a single callback to be attached again. In that case, again, there could be multiple threads racing to do that. The second interlocked operation will guarantee only one of the multiple threads will win.

Reply:
It probably works, but it is getting pretty complicated. Also, the magic numbers like `4` do not look pretty.

Instead of allocating integer IDs to identify the finalizer workitems, we can just allocate a piece of memory that is holding the information about the finalizer work item that is more universal and scalable. Something like this: https://github.com/jkotas/runtime/commit/c97c5e8ea64a631a753895538793cb9d72013fa3 (the delta is untested and for native AOT only). What do you think?

Reply:
I was trying to avoid having the native code to manage the live time of the callback, and I was avoiding having a queue of work items. Your change undoes all of that. That being said, it appears to be simpler than I thought it would be.

I augmented your change with the CoreCLR implementation and made a couple of bug fixes inside the GC. The test cases I introduced are running just fine locally with the changes.